src/Products/PageTemplates/engine.py

```diff
@@ -40,56 +46,142 @@
+
+
+class MappedExprType(object):
+    """map expression type: ``zope.tales`` --> ``chameleon.tales``."""
+    def __init__(self, engine, type):
+        self.engine = engine
+        self.type = type
+
+    def __call__(self, expression):
+        return MappedExpr(self.type, expression, self.engine)
+
+
+class ZtPageTemplate(ChameleonPageTemplate):
+    """``ChameleonPageTemplate`` for use with ``zope.tales`` expressions."""
+
+    # override to get the proper ``zope.tales`` default marker
+    def _compile(self, body, builtins):
+        code = super(ZtPageTemplate, self)._compile(body, builtins)
+        # redefine ``__default`` as ``zope.tales.tales._default``
+        #  Potentially this could be simpler
+        frags = code.split("\n", 5)
+        for i, frag in enumerate(frags[:-1]):
+            if frag.startswith("__default = "):
+                frags[i] = "from zope.tales.tales import _default as __default"
+                break
+        else:
+            raise RuntimeError("unexpected code prelude %s" % frags[:-1])
+        return "\n".join(frags)
+
+
 @implementer(IPageTemplateProgram)
 @provider(IPageTemplateEngine)
 class Program(object):
 
-    # Zope 2 Page Template expressions
-    secure_expression_types = {
-        'python': UntrustedPythonExpr,
-        'string': StringExpr,
-        'not': NotExpr,
-        'exists': ExistsExpr,
-        'path': PathExpr,
-        'provider': ProviderExpr,
-        'nocall': NocallExpr,
-    }
-
-    # Zope 3 Page Template expressions
-    expression_types = {
-        'python': PythonExpr,
-        'string': StringExpr,
-        'not': NotExpr,
-        'exists': ExistsExpr,
-        'path': TrustedPathExpr,
-        'provider': ProviderExpr,
-        'nocall': NocallExpr,
-    }
-
-    extra_builtins = {
-        'modules': ZRPythonExpr._SecureModuleImporter()
-    }
-
-    def __init__(self, template):
+    def __init__(self, template, engine):
         self.template = template
+        self.engine = engine
 
     def __call__(self, context, macros, tal=True, **options):
         if tal is False:
             return self.template.body
 
         # Swap out repeat dictionary for Chameleon implementation
-        # and store wrapped dictionary in new variable -- this is
-        # in turn used by the secure Python expression
-        # implementation whenever a 'repeat' symbol is found
         kwargs = context.vars
-        kwargs['wrapped_repeat'] = kwargs['repeat']
         kwargs['repeat'] = RepeatDict(context.repeat_vars)
+        # provide context for ``zope.tales`` expression compilation
+        #   and evaluation
+        #   unused for ``chameleon.tales`` expressions
+        kwargs["__zt_engine__"] = self.engine
+        kwargs["__zt_context__"] = context
 
         return self.template.render(**kwargs)
 
     @classmethod
     def cook(cls, source_file, text, engine, content_type):
-        if engine is getEngine():
+        if getattr(engine, "untrusted", False):
```

Comment:
The defaults in other places tend to err on the side of caution and assume an untrusted interaction unless there is a specific marker or flag that signals a trusted interaction. Here this is reversed. Is that an oversight?